src/repl/ast/parser/error.rs

```diff
@@ -0,0 +1,76 @@
+use crate::clarity::diagnostic::{DiagnosableError, Level};
+use crate::clarity::representations::Span;
+use crate::repl::ast::parser::lexer::error::LexerError;
+use crate::repl::ast::parser::lexer::token::Token;
+
+#[derive(Debug, PartialEq)]
+pub enum ParserError {
+    // Errors
+    Lexer(LexerError),
+    ContractNameTooLong(String),
+    ExpectedContractIdentifier,
+    ExpectedTraitIdentifier,
+    ExpectedIdentifier,
+    ExpectedPrincipal,
+    IllegalContractName,
+    IllegalTraitName,
+    InvalidPrincipalLiteral,
+    InvalidBuffer,
+    NameTooLong(String),
+    UnexpectedToken(Token),
+    ExpectedClosing(Token),
+    TupleColonExpected,
+    TupleCommaExpected,
+    TupleValueExpected,
+    IllegalClarityName(String),
+    IllegalASCIIString(String),
+    IllegalUtf8String(String),
+    // Notes
+    NoteToMatchThis(Token),
+}
+
+pub struct PlacedError {
+    pub e: ParserError,
+    pub span: Span,
+}
+
+impl DiagnosableError for ParserError {
+    fn message(&self) -> String {
+        use self::ParserError::*;
+        match self {
+            Lexer(le) => le.message(),
+            ContractNameTooLong(name) => format!("contract name, '{}' is too long", name),
+            ExpectedContractIdentifier => "expected contract identifier".to_string(),
+            ExpectedTraitIdentifier => "expected trait identifier".to_string(),
+            ExpectedIdentifier => "expected identifier".to_string(),
+            ExpectedPrincipal => "expected principal".to_string(),
+            IllegalContractName => "illegal contract name".to_string(),
+            IllegalTraitName => "illegal trait name".to_string(),
+            InvalidPrincipalLiteral => "invalid principal literal".to_string(),
+            InvalidBuffer => "invalid hex-string literal".to_string(),
+            NameTooLong(name) => format!("illegal name (too long), '{}'", name),
+            UnexpectedToken(token) => format!("unexpected '{}'", token),
+            ExpectedClosing(token) => format!("expected closing '{}'", token),
+            TupleColonExpected => "expected ':' after key in tuple".to_string(),
+            TupleCommaExpected => "expected ',' separating key-value pairs in tuple".to_string(),
+            TupleValueExpected => "expected value expression for tuple".to_string(),
+            IllegalClarityName(name) => format!("illegal clarity name, '{}'", name),
```

Comment:
Do you think we could revisit the level for this class of issues, and go with a `Warning` or `Info`?
With Clarinet, developers can add `links` to their Clarinet manifest files, which pull the source of the contracts from mainnet / testnet nodes. If we go with an `error`, that will be breaking these projects.

Reply:
Can you give an example of what that looks like? It would be nice if we could still catch and report an error for a truly illegal name and also handle that case.

Reply:
I realized that there is one ugly change needed in order to support the weird names that are currently legal. Currently, the lexer just looks for identifiers, without any context. If the `>` character is allowed in an identifier, then the lexer will need to have context about whether it is looking for a trait-identifier, in which case the `>` might be closing the trait identifier, or any other identifier, in which case that is just part of the name. 

The v1 parser is not able to report an error when defining a trait with a bad name like this, but only gets confused when it is used:
```clarity
(define-trait multipl>ier
    ((multiply (uint uint) (response uint uint)))
)
(define-public (my-multiply (untrusted <multipl>ier>) (a uint) (b uint))
    (contract-call? untrusted multiply a b)
)
```
Old parser output:
```
$ clarinet check
error: Expected whitespace or a close parens. Found: '>'
```

I can solve this pretty easily, but I don't love it.

Reply:
@lgalabru, please take a look at f20e58c with this modification.